src/checked_transaction.rs

```diff
@@ -1,396 +1,340 @@
+//! A checked transaction is type-wrapper for transactions which have been checked.
+//! It is impossible to construct a checked transaction without performing necessary checks.
 //!
-//! This allows the VM to accept transactions that have been already verified upstream,
-//! and consolidates logic around fee calculations and free balances.
+//! This allows the VM to accept transactions with metadata that have been already verified upstream.
 
 use crate::{
-    ConsensusParameters, Input, Metadata, Output, Transaction, TransactionFee, ValidationError,
+    field, Chargeable, CheckError, Checkable, ConsensusParameters, Create, Input, Output, Script,
+    Transaction, TransactionFee,
 };
-use fuel_asm::PanicReason;
-use fuel_types::bytes::SerializableVec;
-use fuel_types::{Address, AssetId, Bytes32, Word};
+use fuel_types::{AssetId, Word};
 
 use alloc::collections::BTreeMap;
-use core::{borrow::Borrow, ops::Index};
-
-use core::mem;
-use std::io::{self, Read};
-
+use core::borrow::Borrow;
+
+/// The type describes that the inner transaction was already checked.
+///
+/// All fields are private, and there is no constructor, so it is impossible to create the instance
+/// of `Checked` outside the `fuel-tx` crate.
+///
+/// The inner data is immutable to prevent modification to invalidate the checking.
+///
+/// If you need to modify an inner state, you need to get inner values
+/// (via the `Into<(Tx, Tx ::Metadata)>` trait), modify them and check again.
+///
+/// Each ascending checked status includes all previous checks. So `Signature` means that
+/// `Stateless` also is passed.
+///
+/// # Dev note: Avoid serde serialization of this type. Since checked tx would need to be
+/// re-validated on deserialization anyways, it's cleaner to redo the tx check.
 #[derive(Debug, Clone, Eq, PartialEq, Hash)]
-// Avoid serde serialization of this type. Since checked tx would need to be re-validated on
-// deserialization anyways, it's cleaner to redo the tx check.
-pub struct CheckedTransaction {
-    /// The transaction that was validated
-    transaction: Transaction,
-    /// The mapping of initial free balances
-    initial_free_balances: BTreeMap<AssetId, Word>,
-    /// The block height this tx was verified with
-    block_height: Word,
-    /// The fees and gas usage
-    fee: TransactionFee,
-    /// Signatures verified
-    checked_signatures: bool,
-}
-
-impl Default for CheckedTransaction {
-    fn default() -> Self {
-        Self::check(Default::default(), Default::default(), &Default::default())
-            .expect("default tx should produce a valid checked transaction")
-    }
+pub enum Checked<Tx: IntoChecked> {
+    /// The transaction passed checks without the usage of the state of the blockchain.
+    ///
+    /// The rules defined in the specification:
+    /// https://github.com/FuelLabs/fuel-specs/blob/master/specs/protocol/tx_format.md#transaction
+    Stateless(Tx, Tx::Metadata),
```

Comment:
Not sure if stateless is the most correct term, since it still requires the block height for validation, whereas signatures are actually stateless to verify.

It's also not clear whether signatures is inclusive of stateless checks or if they are mutually exclusive checked states 🤔 

Ideally, we'd have some kind of `bit-field` that allows us to compose any combination of independent checks together as needed.

Reply:
Let's move with bitmask https://github.com/FuelLabs/fuel-tx/pull/197/commits/2dbc0c05e0da1d928688d1b585191b6e1259d87a=)